src/mappings/contracts.ts

```diff
@@ -246,16 +248,32 @@ export async function nodeContractCanceled(
 
   if (contractID === BigInt(0)) return
 
-  const savedContract = await ctx.store.get(NodeContract, { where: { contractID } })
+  const savedContract = await ctx.store.get(Contract, { where: { contractID } })
   if (!savedContract) return
 
   savedContract.state = ContractState.Deleted
-  await ctx.store.save<NodeContract>(savedContract)
+  await ctx.store.save<Contract>(savedContract)
+
+  let savedPublicIPs: PublicIp[] = await ctx.store.find(PublicIp, { where: { contractId: contractID }, relations: { farm: true } })
+  Promise.all(savedPublicIPs.map(async (ip) => {
+    ip.contractId = null
+    const farm = await ctx.store.get(Farm, { where: { farmID: ip.farm.farmID } })
+    if (farm) {
+      farm.freeIps += 1
+      await ctx.store.save<Farm>(farm)
+    }
+
+    await ctx.store.save<PublicIp>(ip)
+  }))
+
+  if (savedContract.nodeID) {
+    const savedNode = await ctx.store.get(Node, { where: { nodeID: savedContract.nodeID } })
+    if (!savedNode) return
+    savedNode.freeMRU += savedContract.usedMRU || BigInt(0)
+    savedNode.freeSRU += savedContract.usedSRU || BigInt(0)
+    savedNode.freeHRU += savedContract.usedHRU || BigInt(0)
```

Comment:
I keep seeing this pattern in the code where we first read the data in memory, modify it, and then write it back to the database. I’m not an expert on TypeORM, but in general, I think there are better ways to update a field or a few fields in one query. Doing so in multiple queries is less efficient and susceptible to overwrite problems (not atomic).

I never used typeORM but doing a quick search and checking repository API, I believe its safer to use `.update()` or better `.increment()` and `.decrement()`.
Another alternative is doing so using `QueryBuilder`.